PayloadPkg/FirmwareUpdate/CmdFwUpdate.c

```diff
@@ -0,0 +1,161 @@
+  IN  UINT8                         *CapImage,
+  IN  UINT32                        CapImageSize,
+  OUT EFI_FW_MGMT_CAP_IMAGE_HEADER  **ImageHdr
+  );
+
+/*
+ Parse firmware update command data buffer
+
+ Command format:
+ {FLASHDESCLOCK}
+
+  @param[in]  Buffer      Command buffer.
+  @param[in]  BufLength   Command buffer length
+  @param[out] CmdiData    Command data parsed
+*/
+EFI_STATUS
+ParseCommandStringBuf(
+  IN  UINT8       *Buffer,
+  IN  UINTN       BufLength,
+  OUT CMDI_DATA   *CmdiData
+  )
+{
+  CHAR8      *CurrLine;
+  CHAR8      *NextLine;
+  CHAR8      *EndLine;
+  CHAR8      *EndString;
+  UINT32      LineLen;
+  EFI_STATUS  Status;
+
+  Status = EFI_SUCCESS;
+  CurrLine = (CHAR8 *) Buffer;
+
+  while ((CurrLine != NULL) && (CurrLine < (Buffer + BufLength))) {
+    NextLine = GetNextLine (CurrLine, &LineLen);
+    EndLine  = CurrLine + LineLen;
+    CurrLine = TrimLeft (CurrLine);
+
+    if (CurrLine[0] == '{') {
+      // String format found in line
+      CurrLine += 1;
+      EndString = AsciiStrStr(CurrLine, "}");
+
+      if(EndString != NULL) {
+        if(AsciiStrnCmp(CurrLine, FLASH_DESCRIPTOR_LOCK_STR, AsciiStrLen(FLASH_DESCRIPTOR_LOCK_STR)) == 0) {
+          CmdiData->FlashDescLockDataBuf = CurrLine;
+          CmdiData->FlashDescLockDataLen = EndString - CurrLine;
+          CmdiData->CmdRequested        |=  CMDI_TYPE_SPI_DESCRIPTOR_LOCK;
+        }
+      }
+    }
+    CurrLine = NextLine;
+  }
+
+  return Status;
+}
+
+
+/**
+  Main routine for firmware command updates.
+
+  This function has the logic to perform command updates.
+
+  @param[in] CapImage         The pointer to the firmware update capsule image.
+  @param[in] CapImageSize     The size of capsule image in bytes.
+
+  @retval  EFI_SUCCESS      Update successful
+  @retval  other            error status from the update routine
+**/
+EFI_STATUS
+FwCmdUpdateProcess (
+  IN  UINT8     *CapImage,
+  IN  UINTN     CapImageSize
+  )
+{
+  CMDI_DATA                       CmdData;
+  EFI_FW_MGMT_CAP_IMAGE_HEADER    *CapImageHdr;
+  EFI_STATUS                      Status;
+
+  if ((CapImage == NULL) ||  (CapImageSize == 0)){
+    return EFI_INVALID_PARAMETER;
+  }
+
+  // Find the command update region
+  Status = FindImage (FW_UPDATE_COMP_CMD_UPDATE, CapImage, CapImageSize, &CapImageHdr);
+  if ((EFI_ERROR (Status)) || (CapImageHdr == NULL)) {
+    DEBUG((DEBUG_ERROR, "Finding CMDI update failed with Status = %r\n", Status));
+    return Status;
+  }
+
+  ZeroMem (&CmdData, sizeof (CmdData));
+
+  //Parse command data buffer
+  Status = ParseCommandStringBuf ((CHAR8 *)CapImageHdr + sizeof(EFI_FW_MGMT_CAP_IMAGE_HEADER), CapImageHdr->UpdateImageSize, &CmdData);
+  if (EFI_ERROR (Status)) {
+    DEBUG((DEBUG_ERROR, "Error in command parse Status = %r\n", Status));
+    return Status;
+  }
+
+  DEBUG((DEBUG_ERROR, "Commands Requested = 0x%x\n", CmdData.CmdRequested));
+
+  if(CmdData.CmdRequested & CMDI_TYPE_SPI_DESCRIPTOR_LOCK) {

```

Comment:
Is this common for all platform?  If not, this needs to be moved into platform FWU library for command handling.

Reply:
Yes, interface should be same for all platforms.
Parser would populate CmdData.CmdRequested. If this command is in generic, it would be executed from common. If its platform specific (current FlashDescLock), platform FWU library would handle it.